sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -409,6 +409,25 @@ class SessionCatalog(
   }
 
   /**
+   * Returns the partition names from catalog for a given table in a database.
+   */
+  def getPartitionNames(db: String, table: String, range: Short): Seq[String] = {
+    externalCatalog.getPartitionNames(db, table, range)
+  }
+
+  /**
+   * Returns the partition names that matche the partition spec for a given table in a database.
+   * When no match is found, an empty Sequence is returned.
+   */
+  def getPartitionNames(
```

Comment:
I think a better API is:
```
def listPartitions(
    db: String,
    table: String,
    partialSpec: Option[TablePartitionSpec]): Seq[CatalogTablePartition]
```

Note that we already have a `listPartitions`, but it doesn't allow us to match anything. Once we have this we can do the formatting ourselves in the `ShowPartitions` DDL. This leads to a more consistent catalog API and allows us to provide a less awkward implementation in `InMemoryCatalog`.